app/javascript/packages/pwned-passwords/downloader.js

```diff
@@ -0,0 +1,124 @@
+import EventEmitter from 'node:events';
+import PQueue from 'p-queue';
+import PairingHeap from './pairing-heap.js';
+
+/**
+ * @typedef DownloadOptions
+ *
+ * @prop {string} rangeStart Minimum hash prefix for HaveIBeenPwned Range API
+ * @prop {string} rangeEnd Inclusive maximum hash prefix for HaveIBeenPwned Range API
+ * @prop {number} concurrency Number of parallel downloaders to use to retrieve data
+ * @prop {number} maxSize Maximum number of top hashes to retrieve
+ */
+
+/**
+ * @typedef HashPair
+ * @prop {string} hash SHA-1 password hash for common password
+ * @prop {number} prevalence Prevalance count within known breaches
+ */
+
+const API_ROOT = 'https://api.pwnedpasswords.com/range/';
+
+class Downloader extends EventEmitter {
+  /** @type {string} */
+  rangeStart;
+
+  /** @type {string} */
+  rangeEnd;
+
+  /** @type {PQueue} */
+  downloaders;
+
+  /** @type {number} */
+  maxSize;
+
+  /** @type {PairingHeap<HashPair>} */
+  commonHashes = new PairingHeap((a, b) => a.prevalence - b.prevalence);
+
+  /** @type {TextDecoder} */
+  decoder = new TextDecoder();
+
+  /**
+   * @param {Partial<DownloadOptions>} options
+   */
+  constructor({ rangeStart = '00000', rangeEnd = 'fffff', concurrency = 40, maxSize = 3_000_000 }) {
+    super();
+
+    this.rangeStart = rangeStart;
+    this.rangeEnd = rangeEnd;
+    this.maxSize = maxSize;
+    this.downloaders = new PQueue({ concurrency });
+  }
+
+  /**
+   * Downloads the top password hashes from the configured range and resolves with an iterable
+   * object containing all hashes in no particular order.
+   *
+   * @return {Promise<Iterable>}
+   */
+  async download() {
+    const start = parseInt(this.rangeStart, 16);
+    const end = parseInt(this.rangeEnd, 16);
+    const total = end - start + 1;
+    this.emit('start', { total });
+    for (let i = start; i <= end; i++) {
+      this.downloaders.add(async () => {
+        await this.#downloadRange(this.#getRangePath(i));
+        this.emit('download');
+      });
+    }
+
+    await this.downloaders.onIdle();
+    this.emit('complete');
+
+    const { commonHashes } = this;
+    return {
+      *[Symbol.iterator]() {
+        yield* commonHashes;
+      },
+    };
+  }
+
+  /**
+   * Given a number between 0 and 1048575 (16^5 - 1), returns the normalized value to be used as the
+   * path suffix for the HaveIBeenPwned range API (a padded, uppercase base16 number).
+   *
+   * @param {number} value
+   * @return {string}
+   */
+  #getRangePath(value) {
+    return value.toString(16).padStart(5, '0').toUpperCase();
+  }
+
+  /**
+   * Downloads a given range and appends common password hashes from the response.
+   *
+   * @param {string} range
+   */
+  async #downloadRange(range) {
+    const url = new URL(range, API_ROOT);
+    const response = await fetch(url);
+    const text = await response.text();
```

Comment:
I'd give it 3 retries (to start), re-add to queue

Also unsure how it compares, but for ruby I made sure to pull in [net-http-persistent](https://github.com/drbrain/net-http-persistent) to re-use HTTPS connections so it doesn't do the SSL handshake over and over and over again

Reply:
or alternatively if we did the way where it can pick up where it left off based on the FS, I'd do that

Reply:
> I'd give it 3 retries (to start), re-add to queue

I pushed f64c7ee before I saw your comment, but it matches this aside from using 5 vs. 3 retries.

> Also unsure how it compares, but for ruby I made sure to pull in [net-http-persistent](https://github.com/drbrain/net-http-persistent) to re-use HTTPS connections so it doesn't do the SSL handshake over and over and over again

I looked into this a bit more and my understanding is that Node can internally manage this for us, though it may need to be configured. Since we're using the `fetch` global abstraction, I'll take a look at how this can be configured.

Specifically: https://nodejs.org/api/http.html#class-httpagent

>It maintains a queue of pending requests for a given host and port, reusing a single socket connection for each until the queue is empty, at which time the socket is either destroyed or put into a pool where it is kept to be used again for requests to the same host and port. Whether it is destroyed or pooled depends on the keepAlive [option](https://nodejs.org/api/http.html#new-agentoptions).

Strangely, despite [being documented as `false` by default](https://nodejs.org/api/http.html#new-agentoptions), the global agent appears to be set to `true` in my local testing 🤔 

```
$ node -p "require('https').globalAgent.keepAlive"
true
```

Reply:
It's not documented anywhere, but it appears the global agent is initialized as `keepAlive` by default:

https://github.com/nodejs/node/blob/c1fa929f3967444fdf29db24d9c63cc13ff92bb6/lib/https.js#L354

Not entirely clear yet that `fetch` uses the `https` module / `globalAgent` internally, but I suspect it does.

Reply:
I had trouble pinning down whether Node's global fetch internally uses the same `https` mechanism, but also wondered if reverting to using plain `https` would have a few advantages:

- Gives us more guarantees that connection pooling works
- Related to https://github.com/18F/identity-idp/pull/10108#discussion_r1497692796, reading a response with `https` works very similar to the old approach with using the body reader, and may not have the same conflicts with `msw`
- It may not throw the same errors I was seeing with `fetch` (best case scenario, we may not need the retry at all?)